src/main/java/com/minecolonies/coremod/entity/ai/citizen/builder/EntityAIStructureBuilder.java

```diff
@@ -128,15 +132,55 @@ public IAIState getStateAfterPickUp()
 
     /**
      * State to pick up material before going back to work.
+     *
      * @return the next state to go to.
      */
     public IAIState pickUpMaterial()
     {
         final BuildingBuilder building = getOwnBuilding();
-        final List<Predicate<ItemStack>> neededItemsList = new ArrayList<>();
-        for (final BuildingBuilderResource stack : building.getNeededResources().values())
+        final List<Tuple<Predicate<ItemStack>, Integer>> neededItemsList = new ArrayList<>();
+        boolean ONLY_64 = false;
+
+        Map<String, BuildingBuilderResource> neededRessourcesMap = building.getNeededResources();
+        if (neededRessourcesMap.size() > InventoryUtils.openSlotCount(worker.getInventoryCitizen()) - MIN_OPEN_SLOTS)
         {
-            neededItemsList.add(itemstack -> ItemStackUtils.compareItemStacksIgnoreStackSize(stack.getItemStack(), itemstack, true, true));
+            ONLY_64 = true;
+        }
+        else
+        {
+            int stackCount = 0;
```

Comment:
This need to be reworked. The logic here does not take stackability of items which are already contained in the inventory of the worker into account. Also formatting of while loops is of.

Reply:
Resolved